---
title: "Part 2 TP NGS single cell - Making the dental cell atlas"
output:
  html_document: default
  word_document: default
  pdf_document: default
---


```{r tximport, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("tximport")
setwd (ifb/data/mydatalocal/data)
dir <- "/ifb/data/mydatalocal/data/alignment/quant_results"

files=list.files(dir)
cells <- gsub(".fastq.gz","",files)
files1=paste0(dir,"/",files,"/","quant.sf")
names(files1)=cells

library(biomaRt)

ensembl <- useEnsembl(biomart = "genes", dataset = "mmusculus_gene_ensembl")
attributeNames <- c('ensembl_transcript_id','external_gene_name')
annot <- getBM(attributes=attributeNames, 
               mart = ensembl)
names(annot) <- c("txname","geneid")

txi1 <- tximport(files1[1:500],type="salmon",tx2gene=annot,ignoreTxVersion=T)

txi2 <- tximport(files1[501:1000],type="salmon",tx2gene=annot,ignoreTxVersion=T)
txi3 <- tximport(files1[1001:1500],type="salmon",tx2gene=annot,ignoreTxVersion=T)
txi4 <- tximport(files1[1501:2000],type="salmon",tx2gene=annot,ignoreTxVersion=T)
txi5 <- tximport(files1[2001:2553],type="salmon",tx2gene=annot,ignoreTxVersion=T)

txi <- txi1
txi$"abundance" <- cbind(txi1$"abundance",txi2$"abundance",txi3$"abundance",txi4$"abundance",txi5$"abundance")
txi$"counts" <- cbind(txi1$"counts",txi2$"counts",txi3$"counts",txi4$"counts",txi5$"counts")
txi$"length" <- cbind(txi1$"length",txi2$"length",txi3$"length",txi4$"length",txi5$"length")
txi$"countsFromAbundance" <- cbind(txi1$"countsFromAbundance",txi2$"countsFromAbundance",txi3$"countsFromAbundance",txi4$"countsFromAbundance",txi5$"countsFromAbundance")

saveRDS(txi, "txi.rds")

```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r tximport2}
 txi2=readRDS("txi.rds")

```

## Including Plots

You can also embed plots, for example:

```{r Seurat}
library(dplyr)
library(Seurat)
setwd("/ifb/data/mydatalocal/tp_ngs_single_cell")

```


```{r Seurat, eval=FALSE}

incisor <- CreateSeuratObject(counts = txi$counts, project = "incisor", min.cells = 3, min.features = 200)
incisor

```


```{r ADNmt, eval=FALSE}
incisor[["percent.mt"]] <- PercentageFeatureSet(incisor, pattern = "^mt-")
# Visualize QC metrics as a violin plot
VlnPlot(incisor, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```


```{r filter, eval=FALSE}
plot1 <- FeatureScatter(incisor, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(incisor, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

#Technique de filtrage des cellules 1
#incisor <- subset(incisor, subset = nFeature_RNA > 500 & nFeature_RNA < 10000 & percent.mt < 15 & nCount_RNA < 1000000)

#Technique de filtrage des cellules 2
quantile5_incisor <- quantile(incisor$nFeature_RNA, probs = c(0.05), na.rm = FALSE, names = TRUE, type = 7)
incisor <- subset(incisor, subset = nFeature_RNA > quantile5_incisor & percent.mt < 15 & nCount_RNA < 1000000)

VlnPlot(incisor, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r Normalization, eval=FALSE}
incisor <- NormalizeData(incisor, normalization.method = "LogNormalize", scale.factor = 10000)
```

```{r Variant identification, eval=FALSE}
#Apply the selection method to detect differentially expressed genes (On a 28851 transcrits différents -> quantile à 90%)

incisor <- FindVariableFeatures(incisor, selection.method = "vst", nfeatures = 2885)
VariableFeatures(incisor)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(incisor), 10)
top10

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(incisor)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE, xnudge=0, ynudge=0)
plot1
plot2
```

```{r Data Scaling, eval=FALSE}
#Sélectionne tous les noms de gènes
all.genes <- rownames(incisor)
incisor <- ScaleData(incisor, features = all.genes)
```

```{r PCA, eval=FALSE}

incisor <- RunPCA(incisor, features = VariableFeatures(object = incisor))

print(incisor[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(incisor, dims = 1:2, reduction = "pca")

DimPlot(incisor, reduction = "pca")

DimHeatmap(incisor, dims = 1, cells = 2257, balanced = TRUE)

DimHeatmap(incisor, dims = 1:15, cells = 2257, balanced = TRUE)
```

``` {r Dimensionality, eval=FALSE}
incisor <- JackStraw(incisor, num.replicate = 100)
incisor <- ScoreJackStraw(incisor, dims = 1:20)
JackStrawPlot(incisor, dims = 1:20)
ElbowPlot(incisor)

```

``` {r Clustering, eval=FALSE}
incisor <- FindNeighbors(incisor, dims = 1:20)
incisor <- FindClusters(incisor, resolution = 0.5)
head(Idents(incisor), 5)
```

```{r UMAP/tSNE, eval=FALSE}
incisor <- RunUMAP(incisor, dims = 1:20)
DimPlot(incisor, reduction = "umap")
saveRDS(incisor, file = "incisor.rds")
```

```{r Cluster biomarkers, eval=FALSE}
cluster1.markers <- FindMarkers(incisor, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 5)

 incisor.markers <- FindAllMarkers(incisor, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

incisor.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)

# Ajout du ROC test qui donne le 'classification power' de chaque marqueur, compris entre 0 et 1
cluster1.markers <- FindMarkers(incisor, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
head(cluster1.markers, n = 5)

VlnPlot(incisor, features = c("Mrc1", "C1qb"))
VlnPlot(incisor, features = c("Mrc1", "C1qb"), slot = "counts", log = TRUE)

incisor.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
FeaturePlot(incisor, features = c("Smoc2", "C1qb","Tnc","Emcn","Car2","Sfrp2","Cd83","Epcam"))
FeaturePlot(incisor, features = c("S100a9","Cd209a","Ibsp","Scn7a","Rgs5","Krt17","Gm10801","Gm17660"))

top10 <- incisor.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(incisor, features = top10$gene) + NoLegend()

#Avec moins de gènes par cluster pour plus de lisibilité de la HeatMap
top2 <- incisor.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
DoHeatmap(incisor, features = top2$gene) + NoLegend()

```
