---
title: "Untitled"
output:
  html_document: default
  word_document: default
  pdf_document: default
---

```{r tximport, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("tximport")
setwd (ifb/data/mydatalocal/data)
dir <- "/ifb/data/mydatalocal/data/alignment/quant_results"

files=list.files(dir)
cells <- gsub(".fastq.gz","",files)
files1=paste0(dir,"/",files,"/","quant.sf")
names(files1)=cells

library(biomaRt)

ensembl <- useEnsembl(biomart = "genes", dataset = "mmusculus_gene_ensembl")
attributeNames <- c('ensembl_transcript_id','external_gene_name')
annot <- getBM(attributes=attributeNames, 
               mart = ensembl)
names(annot) <- c("txname","geneid")

txi1 <- tximport(files1[1:500],type="salmon",tx2gene=annot,ignoreTxVersion=T)

txi2 <- tximport(files1[501:1000],type="salmon",tx2gene=annot,ignoreTxVersion=T)
txi3 <- tximport(files1[1001:1500],type="salmon",tx2gene=annot,ignoreTxVersion=T)
txi4 <- tximport(files1[1501:2000],type="salmon",tx2gene=annot,ignoreTxVersion=T)
txi5 <- tximport(files1[2001:2553],type="salmon",tx2gene=annot,ignoreTxVersion=T)

txi <- txi1
txi$"abundance" <- cbind(txi1$"abundance",txi2$"abundance",txi3$"abundance",txi4$"abundance",txi5$"abundance")
txi$"counts" <- cbind(txi1$"counts",txi2$"counts",txi3$"counts",txi4$"counts",txi5$"counts")
txi$"length" <- cbind(txi1$"length",txi2$"length",txi3$"length",txi4$"length",txi5$"length")
txi$"countsFromAbundance" <- cbind(txi1$"countsFromAbundance",txi2$"countsFromAbundance",txi3$"countsFromAbundance",txi4$"countsFromAbundance",txi5$"countsFromAbundance")

saveRDS(txi, "txi.rds")

```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r tximport2}
 txi2=readRDS("txi.rds")

```

## Including Plots

You can also embed plots, for example:

```{r Seurat}
library(dplyr)
library(Seurat)
setwd("/ifb/data/mydatalocal/tp_ngs_single_cell")

```


```{r Seurat, eval=FALSE}

incisor <- CreateSeuratObject(counts = txi$counts, project = "incisor", min.cells = 3, min.features = 200)
incisor

```


```{r ADNmt, eval=FALSE}
incisor[["percent.mt"]] <- PercentageFeatureSet(incisor, pattern = "^mt-")
# Visualize QC metrics as a violin plot
VlnPlot(incisor, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```


```{r filter, eval=FALSE}
plot1 <- FeatureScatter(incisor, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(incisor, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

incisor <- subset(incisor, subset = nFeature_RNA > 500 & nFeature_RNA < 10000 & percent.mt < 15 & nCount_RNA < 1000000)
VlnPlot(incisor, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r Normalization, eval=FALSE}
incisor <- NormalizeData(incisor, normalization.method = "LogNormalize", scale.factor = 10000)
```

```{r Variant identification, eval=FALSE}
#Apply the selection method to detect differentially expressed genes
incisor <- FindVariableFeatures(incisor, selection.method = "vst", nfeatures = 2000)
VariableFeatures(incisor)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(incisor), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(incisor)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE, xnudge=0, ynudge=0)
plot1
plot2
```

```{r Data Scaling, eval=FALSE}
#Sélectionne tous les noms de gènes
all.genes <- rownames(incisor)
incisor <- ScaleData(incisor, features = all.genes)
```

```{r PCA, eval=FALSE}

incisor <- RunPCA(incisor, features = VariableFeatures(object = incisor))

print(incisor[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(incisor, dims = 1:2, reduction = "pca")

DimPlot(incisor, reduction = "pca")

DimHeatmap(incisor, dims = 1, cells = 500, balanced = TRUE)

DimHeatmap(incisor, dims = 1:15, cells = 500, balanced = TRUE)
```

``` {r PCA, eval=FALSE}
incisor <- JackStraw(incisor, num.replicate = 100)
incisor <- ScoreJackStraw(incisor, dims = 1:20)
JackStrawPlot(incisor, dims = 1:20)

```
